libgadget/sfr_eff.c

```diff
@@ -23,373 +23,118 @@
-        TreeWalk tw[1] = {{0}};
-
-        tw->ev_label = "SFR_WIND";
-        tw->fill = (TreeWalkFillQueryFunction) sfr_wind_copy;
-        tw->reduce = (TreeWalkReduceResultFunction) sfr_wind_reduce_weight;
-        tw->UseNodeList = 1;
-        tw->query_type_elsize = sizeof(TreeWalkQueryWind);
-        tw->result_type_elsize = sizeof(TreeWalkResultWind);
-
-        /* sum the total weight of surrounding gas */
-        tw->ngbiter_type_elsize = sizeof(TreeWalkNgbIterWind);
-        tw->ngbiter = (TreeWalkNgbIterFunction) sfr_wind_weight_ngbiter;
-
-        /* First set DensityIterationDone for weighting */
-        /* Watchout: the process function name is preprocess, but not called in the feedback tree walk
-         * because we need to compute the normalization before the feedback . */
-        tw->visit = NULL;
-        tw->haswork = (TreeWalkHasWorkFunction) sfr_wind_feedback_haswork;
-        tw->postprocess = (TreeWalkProcessFunction) sfr_wind_feedback_preprocess;
-        treewalk_run(tw, ActiveParticle, NumActiveParticle);
-
-        tw->haswork = sfr_wind_weight_haswork;
-        tw->visit = (TreeWalkVisitFunction) treewalk_visit_ngbiter;
-        tw->postprocess = (TreeWalkProcessFunction) sfr_wind_weight_postprocess;
-
-        int done = 0;
-        while(!done) {
-            NPLeft = 0;
-            treewalk_run(tw, ActiveParticle, NumActiveParticle);
-
-            int64_t totalleft = 0;
-            sumup_large_ints(1, &NPLeft, &totalleft);
-            message(0, "Star DM iteration Total left = %ld\n", totalleft);
-            done = totalleft == 0;
-        }
-
-        /* Then run feedback */
-        tw->haswork = (TreeWalkHasWorkFunction) sfr_wind_feedback_haswork;
-        tw->ngbiter = (TreeWalkNgbIterFunction) sfr_wind_feedback_ngbiter;
-        tw->postprocess = (TreeWalkProcessFunction) sfr_wind_feedback_postprocess;
-        tw->reduce = NULL;
-
-        treewalk_run(tw, ActiveParticle, NumActiveParticle);
-        myfree(Wind);
-    }
-    walltime_measure("/Cooling/Wind");
-}
-
-#else //No SFR
-
-static void
-cool_postprocess(int i, TreeWalk * tw)
-{
-    cooling_direct(i);
-}
-
-/* cooling routine when star formation is disabled */
-void cooling_only(void)
-{
-    if(!All.CoolingOn) return;
-    walltime_measure("/Misc");
-
-    TreeWalk tw[1] = {0};
-
-    /* Only used to list all active particles for the parallel loop */
-    /* no tree walking and no need to export / copy particles. */
-
-    tw->visit = NULL; /* no tree walk */
-    tw->ev_label = "SFR_COOL";
-    tw->haswork = sfr_cooling_haswork;
-    tw->postprocess = (TreeWalkProcessFunction) cool_postprocess;
-
-    treewalk_run(tw, ActiveParticle, NumActiveParticle);
-
-    walltime_measure("/Cooling/StarFormation");
-}
-
+#ifdef SFR
+    /* Now apply the wind model: has to use the new NumActiveParticle.*/
+    winds_and_feedback(ActiveParticle, NumActiveParticle);
```

Comment:
But NumActiveParticle is also a global variable?

Reply:
Comment is badly worded: I mean that NumActiveParticle changes within the loop when new particles are spawned, so one cannot use nactive.